### PR TITLE
Handle network failures when connection requests cross.

### LINKED ISSUE #23: [NTTCP-9] Connection crossing can block programs indefinitely
[Imported from JIRA. Reported by Facundo Dominguez @facundominguez)  as [NTTCP-9](https://cloud-haskell.atlassian.net/browse/NTTCP-9) on 2015-05-08 21:30:44]
The [comments|https://github.com/haskell-distributed/network-transport-tcp/blob/773ca26f69d79226e401665f7e6589d5c7f00977/src/Network/Transport/TCP.hs#L159-L181] in the code explain how crossing connections are resolved.

This solution, unfortunately, does not account for the possibility of connection failures.

It can be that A has a severed connection to B, but A still thinks the connection is healthy. Now, B is aware that the connection has failed, and it tries to reconnect. The result is that A will respond with {{ConnectionRequestCrossed}}, and B will wait forever an incoming connection from A.

### PR DESCRIPTION
**Pull request by facundominguez**
Fixes #23 NTTCP-9

The protocol for resolving crossed connections was not able to deal with connection failures.

This patch implements a probing mechanism which allows the transport to judge if a connection is healthy and discards it if not, as described in https://cloud-haskell.atlassian.net/browse/NTTCP-9?focusedCommentId=18801&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-18801.

On the client side, remote endpoints in init state are dropped when timeouts occur while waiting for the remote side to establish a connection.

### CODE REVIEW

File: src/Network/Transport/TCP.hs [hunk 1]
```diff
@@ -830,6 +849,36 @@ handleConnectionRequest transport sock = handle handleException $ do
...
+              -- ack does not arrive on time).
+              void $ forkIO $ do
+                tryCloseSocket (remoteSocket vst)
+                let closed = RemoteEndPointFailed $ userError "socket failed"
+                modifyMVar_ (remoteState theirEndPoint) $ \st' -> case st' of
+                  RemoteEndPointValid _ -> return closed
+                  RemoteEndPointClosing _ _ -> return closed
+                  _ -> return st'
+                removeRemoteEndPoint (ourEndPoint, theirEndPoint)
+                void $ tryIO $ createRemoteEndPoint params ourEndPoint
```
Comment (facundominguez) [inline src/Network/Transport/TCP.hs]: Maybe reconnecting is not worth the complexity here. The other side will probably timeout before we try to reconnect.

Comment (facundominguez) [inline src/Network/Transport/TCP.hs]: Also, this is not communicating the network failure to upper layers.

File: src/Network/Transport/TCP.hs [hunk 2]
```diff
@@ -737,9 +750,13 @@ apiCloseEndPoint transport evs ourEndPoint =
               void $ tryIO $ sendOn vst [ encodeInt32 CloseSocket
                                         , encodeInt32 (vst ^. remoteMaxIncoming)
                                         ]
+              -- Release probing resources if probing.
+              forM_ (remoteProbing vst) $ id
```
Comment (qnikst) [inline src/Network/Transport/TCP.hs]: Redundant `$`

File: src/Network/Transport/TCP.hs [hunk 3]
```diff
@@ -737,9 +750,13 @@ apiCloseEndPoint transport evs ourEndPoint =
...
               tryCloseSocket (remoteSocket vst)
             return closed
           RemoteEndPointClosing resolved vst -> do
+            -- Release probing resources if probing.
+            forM_ (remoteProbing vst) $ id
```
Comment (qnikst) [inline src/Network/Transport/TCP.hs]: same here

File: src/Network/Transport/TCP.hs [hunk 4]
```diff
@@ -830,6 +849,34 @@ handleConnectionRequest transport sock = handle handleException $ do
...
+    probeIfValid :: RemoteEndPoint -> IO ()
+    probeIfValid theirEndPoint = modifyMVar_ (remoteState theirEndPoint) $
+      \st -> case st of
+        RemoteEndPointValid
+          vst@(ValidRemoteEndPointState { remoteProbing = Nothing }) -> do
+            tid <- forkIO $ do
+              -- send probe
+              let params = transportParams transport
+              void $ tryIO $ System.Timeout.timeout
+                  (maybe (-1) id $ transportConnectTimeout params) $ do
```
Comment (qnikst) [inline src/Network/Transport/TCP.hs]: `fromMaybe (-1) $ transportConnectTimeout params` ?

**Comment by qnikst** [comment]
Modulo cosmetic comments looks good to me. I don't see failure cases or races in current implementation.

**Pull request merged**